src/Pos/Wallet/Web/Tracking.hs

```diff
@@ -192,8 +200,40 @@ syncWalletsWithGState
       Ether.MonadReader' GenesisUtxo m)
     => [EncryptedSecretKey]
     -> m ()
-syncWalletsWithGState encSKs = withBlkSemaphore_ $ \tip ->
-    tip <$ mapM_ (syncWalletWithGStateUnsafe @ssc) encSKs
+syncWalletsWithGState encSKs = forM_ encSKs $ \encSK -> do
+    let wAddr = encToCId encSK
+    whenJustM (WS.getWalletSyncTip wAddr) $ \wTip ->
+        if wTip == genesisHash then syncDo encSK Nothing
+        else DB.blkGetHeader wTip >>= \case
+            Nothing ->
+                logWarning $
+                sformat ("Couldn't get block header of wallet "%build
+                            %" by last synced hh: "%build) wAddr wTip
+            Just wHeader -> syncDo encSK (Just wHeader)
+  where
+    syncDo :: EncryptedSecretKey -> Maybe (BlockHeader ssc) -> m ()
+    syncDo encSK wTipH = do
+        let wdiff = maybe (0::Word32) (fromIntegral . ( ^. difficultyL)) wTipH
+        gstateTipH <- DB.getTipHeader @(Block ssc)
+        -- If account's syncTip is before the current gstate's tip,
+        -- then it loads accounts and addresses starting with @wHeader@.
+        -- syncTip can be before gstate's the current tip
+        -- when we call @syncWalletSetWithTip@ at the first time
+        -- or if the application was interrupted during rollback.
+        -- We don't load all blocks explicitly, because blockain can be long.
+        when (gstateTipH ^. difficultyL > blkSecurityParam + fromIntegral wdiff) $ do
+            -- Wallet tip is "far" from gState tip,
+            -- rollback can't occur more then @blkSecurityParam@ blocks,
+            -- so we can sync wallet and GState without the block lock
+            -- to avoid blocking of blocks verification/application.
+            bh <- unsafeLast . getNewestFirst <$> DB.loadHeadersByDepth (blkSecurityParam + 1) (headerHash gstateTipH)
+            logDebug $
+                sformat ("Wallet's tip is far from GState tip. Syncing with "%build%" without the block lock")
+                (headerHash bh)
+            syncWalletWithGStateUnsafe encSK wTipH bh
+        withBlkSemaphore_ $ \tip -> do
+            tipH <- maybe (error "Wallet tracking: no block header corresponding to tip") pure =<< DB.blkGetHeader tip
+            tip <$ syncWalletWithGStateUnsafe encSK wTipH tipH
```

Comment:
It seems like here you should start from `bh`, not from `wTipH`, in case when previous section of code is executed. Fix, or explain why I'm wrong.